core/src/main/scala/org/trustedanalytics/sparktk/frame/internal/ops/topk/TopKRddFunctions.scala

```diff
@@ -0,0 +1,122 @@
+   * @return Top (or bottom) K distinct values by count for specified column
+   */
+  def topK(frameRdd: RDD[Row], dataColumnIndex: Int, k: Int, useBottomK: Boolean = false,
+           weightsColumnIndexOption: Option[Int] = None,
+           weightsTypeOption: Option[DataType] = None): RDD[Row] = {
+    require(dataColumnIndex >= 0, "label column index must be greater than or equal to zero")
+
+    val dataWeightPairs =
+      ColumnStatistics.getDataWeightPairs(dataColumnIndex, weightsColumnIndexOption, weightsTypeOption, frameRdd)
+        .filter({ case (data, weight) => NumericValidationUtils.isFinitePositive(weight) })
+
+    val distinctCountRDD = dataWeightPairs.reduceByKey((a, b) => a + b)
+
+    //Sort by descending order to get top K
+    val isDescendingSort = !useBottomK
+
+    // Efficiently get the top (or bottom) K entries by first sorting the top (or bottom) K entries in each partition
+    // This function uses a tree map instead of a bounded priority queue (despite the added overhead)
+    // because we need to keep key-value pairs
+    val topKByPartition = distinctCountRDD.mapPartitions(countIterator => {
+      Iterator(sortTopKByValue(countIterator, k, isDescendingSort))
+    }).reduce({ (topPartition1, topPartition2) =>
+      mergeTopKSortedSeqs(topPartition1, topPartition2, isDescendingSort, k)
+    })
+
+    // Get the overall top (or bottom) K entries from partitions
+    // Works when K*num_partitions fits in memory of single machine.
+    val topRows = topKByPartition.map(f => Row(f.key, f.value))
+    frameRdd.sparkContext.parallelize(topRows)
+  }
+
+  /**
+   * Returns top K entries sorted by value.
+   *
+   * The sort ordering may either be ascending or descending.
+   *
+   * @param inputIterator Iterator of key-value pairs to sort
+   * @param k Number of top sorted entries to return
+   * @param descending Sort in descending order if true, else sort in ascending order
+   * @return Top K sorted entries
+   */
+  def sortTopKByValue(inputIterator: Iterator[(Any, Double)],
+                      k: Int, descending: Boolean = false): Seq[CountPair] = {
+    val ordering = if (descending) Ordering[CountPair].reverse else Ordering[CountPair]
+    val priorityQueue = new PriorityQueue[CountPair]()(ordering)
+
+    inputIterator.foreach(element => {
+      priorityQueue.enqueue(CountPair(element._1, element._2))
+      if (priorityQueue.size > k) priorityQueue.dequeue()
+    })
+
+    priorityQueue.reverse.dequeueAll
+  }
+
+  /**
+   * Merge two sorted sequences while maintaining sort order, and return topK.
+   *
+   * @param sortedSeq1 First sorted sequence
+   * @param sortedSeq2 Second sorted sequence
+   * @param descending Sort in descending order if true, else sort in ascending order
+   * @param k Number of top sorted entries to return
+   * @return Merged sorted sequence with topK entries
+   */
+  private def mergeTopKSortedSeqs(sortedSeq1: Seq[CountPair], sortedSeq2: Seq[CountPair],
```

Comment:
prefer  ``sortedSeqA`` and ``sortedSeqB``  esp. since scala idiomatically uses number suffixes to indicate the number of parameters  (like no. of type parameters)